### ISSUE TITLE
Allow roles to raise an error on method name collisions

### ISSUE THREAD

**Issue by saturnflyer** [saturnflyer/surrounded#24]
Contexts should provide an option to raise errors when objects have methods that match the names of the other roles in the context. #23 has comments about related problems

**Comment by wjdhamilton** [comment]
+1 
This feels fairly straightforward. Could you simply output a warning to the console if the object respond_to a role name?

**Comment by saturnflyer** [comment]
I think that the options should be 1) do nothing, 2) warn to a log, and 3) raise an exception.
I played a bit with the code for this but didn't get anything done with it yet

**Comment by wjdhamilton** [comment]
I agree, I think that it ought to be possible to set the option globally for a given application. I was going to suggest that you allow the option to be overriden at the context level, but I can't think why you'd want to do that...

**Comment by wjdhamilton** [comment]
Planning on having a look at this over the weekend, though it won't be until later on Sunday.

**Comment by saturnflyer** [comment]
:+1: thanks.
If I clean up my hacks before then I'll throw up a branch of what I've got that you can use or not.

I'd been playing with the API like this:
```ruby
on_name_collision :raise # raise errors on initialization, but aggregate all errors into 1 (2 objects collide? provide info on both objects, not just the first collision)
on_name_collision :warn # same behavior but just send the message to STDERR
```
My initial thought was that the default behavior would be to do nothing, but perhaps there's going to be cases where one might setup a general class for contexts and want to turn off the warning with `on_name_collision :nothing`...
But that's probably a strange thing to do. I suspect most people will want it on or off.

In general, I think things should be as flexible as any user of the library wants. I prefer to allow context classes to set their requirements individually. It's pretty simple to make a general class which sets things up the way you want and then inherit from that. That, to me, is clearer than having global configs.

Currently, the only global config option is `default_role_type` so this change might push me to remove that and just suggest that people create their own global classes to handle global configuration.
One option here is to add a config where we define `Surrounded::Context.on_name_collision` but as the global config options grow, the library will become harder to understand and use.

One thing to consider (that I'm just thinking of now that I'm writing this) is that the argument could be a method name into which we pass the exception. That would allow anyone to configure their contexts to do whatever crazy thing they want: `on_name_collision :tweet`

**Comment by wjdhamilton** [comment]
I think that generally all makes a lot of sense, particularly your point about using inheritance instead of a global variable. My only comment is that I would prefer the "warn" behaviour to be on by default in a development or test environment, though I appreciate that is probably a job for framework specific versions like Surrounded-rails rather than Surrounded. We could print something to the console like "``MyContext method foo has a name collision with actor bar. To turn this warning off set on_name_collision :off``" or similar. Anyway, today's task is to work out how to detect the name collisions...

**Comment by wjdhamilton** [comment]
Just while we're here, the test suite throws quite a few deprecation warnings around keyword initialisation. Would you like me to fix them?

**Comment by saturnflyer** [comment]
@wjdhamilton if you want to tackle the deprecation warnings, that'd be great! The tests just need to switch to initialize_without_keywords or to initialize_with_keywords and then fix all the arguments. If you do, please keep it separate from the name collisions stuff

**Comment by wjdhamilton** [comment]
No problem. I'm nearly there with the name collisions, just run into a last minute headache trying to set an instance variable using a class method, but I'll work it out. I should have something for you to look at later in the week.

**Comment by wjdhamilton** [comment]
Hi,

Did you notice the pull request I put in for this?

**Comment by saturnflyer** [comment]
Yes! Thanks. Haven't had time to sit down and review. I really appreciate your contribution. I will spend some time today on it

**Comment by wjdhamilton** [comment]
It's no problem, I just haven't done this before so I wasn't sure if you got an eMail or if I had to tell you!

**Comment by saturnflyer** [comment]
closing. implemented with https://github.com/saturnflyer/surrounded/pull/25 and changes merged into d07c483515ff032f6e33ebe78532719de62e891d
